frame/pablo/src/lib.rs

```diff
@@ -686,16 +707,28 @@ pub mod pallet {
 				quote_price_cumulative != T::Balance::zero()
 			{
 				// update TWAP
-				let _ = TWAPState::<T>::try_mutate(
+				let updated_twap = TWAPState::<T>::try_mutate(
 					pool_id,
-					|prev_twap_state| -> Result<(), DispatchError> {
+					|prev_twap_state| -> Result<Option<TWAPStateOf<T>>, DispatchError> {
 						update_twap_state::<T>(
 							base_price_cumulative,
 							quote_price_cumulative,
 							prev_twap_state,
 						)
+						.map_or_else(|_| Ok(None), |_| Ok(prev_twap_state.clone()))
 					},
-				);
+				)?;
+				if let Some(updated_twap) = updated_twap {
+					Self::deposit_event(Event::<T>::TWAPUpdated {
+						pool_id,
+						timestamp: updated_twap.timestamp,
+						twaps: BTreeMap::from([
```

Comment:
you do not need tree here. building tree does hashing and burns resources. you may just use vec.

Reply:
because anyway it was tree and no need to prove that again

Reply:
To support more than two asset I am advised to use tree with asset_id -> balance mapping instead of using vec where we need to assume which balance for which asset. 
For performance I think we can benchmark and see how much difference is there due to hashing and tree creation.

Reply:
IIRC there is no hash map available in substrate std.

Reply:
Correct: https://docs.rs/sp-std/latest/sp_std/collections/index.html

Reply:
insert into btree is O(logN) insert into vec is O(1). why use tree and making log N in hook? this is attach vector.

Reply:
As this is basically an internal operation and the number of TWAP enabled pools are limited, the hook is already O(n) so additional `O(logN)` is still O(n). If there is a way to get rid of O(n) we should definitely do it though like my suggestion on the other thread for event emitting

Reply:
uniseap had 1000 plus pools, there contracta with 3000 pools.

Reply:
i guess comlosable want the same.

Reply:
i do bot get your reasoning

Reply:
TWAP is only for certain pools we enable it for. So all pools won't have twap enabled

Reply:
But adding a comment about these technical considerations in the code makes sense btw

Reply:
until hash is consumed in memory it useles. so now we are doing useless job on chain. instead of doing simple vec. i do not get that logic.

Reply:
what comments? when there is not map user there is not need for map.

Reply:
Map will be used by code which processes event, to identify which is base_amount and which is quote_amount.
Also using map vs vec is yet to be finalized with https://app.clickup.com/t/2gh7cjb
so perhaps after that we may not use map and use vec all the places.